storagemarket/impl/provider.go

```diff
@@ -298,6 +303,53 @@ func (p *Provider) UniversalRetrievalEnabled() bool {
 	return p.universalRetrievalEnabled
 }
 
+func (p *Provider) SubscribeToEvents(subscriber storagemarket.ProviderSubscriber) shared.Unsubscribe {
+	return shared.Unsubscribe(p.pubSub.Subscribe(subscriber))
+}
+
+// dispatch puts the fsm event into a form that pubSub can consume,
+// then publishes the event
+func (p *Provider) dispatch(eventName fsm.EventName, deal fsm.StateType) {
+	evt, ok := eventName.(storagemarket.ProviderEvent)
+	if !ok {
+		log.Errorf("dropped bad event %s", eventName)
+	}
+	realDeal, ok := deal.(storagemarket.MinerDeal)
+	if !ok {
+		log.Errorf("not a deal %v", deal)
+	}
+	pubSubEvt := internalEvent{evt, &realDeal}
+
+	go func() {
```

Comment:
Without this the integration test fails when there is a subscriber, so it seems pubsub is causing a deadlock

Reply:
this is not correct. Remove the go routine, add a buffer one on the dealchan.

It locks because the send to the dealChan locks in the integration test. Channel writes block without a buffer on the channel.

Reply:
that alone will probably not fix it, cause you are writing to the deal chan each time you update state. you either need to make sure you're seperately clearing the deal chan every time, or you need to make sure to write to it only on certain states and then clear it. the listener should NOT be a blocking operation.